File: rb/test/npm_package_entity_test.rb
# NpmPackage entity test

require "minitest/autorun"
require "json"
require_relative "../Gitlab_sdk"
require_relative "runner"

class NpmPackageEntityTest < Minitest::Test
  def test_create_instance
    testsdk = GitlabSDK.test(nil, nil)
    ent = testsdk.NpmPackage(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = npm_package_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "npm_package." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set GITLAB_TEST_NPM_PACKAGE_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    npm_package_ref01_ent = client.NpmPackage(nil)
    npm_package_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.npm_package"), "npm_package_ref01"))
    npm_package_ref01_data["group_id"] = setup[:idmap]["group01"]
    npm_package_ref01_data["project_id"] = setup[:idmap]["project01"]

    npm_package_ref01_data_result = npm_package_ref01_ent.create(npm_package_ref01_data, nil)
    npm_package_ref01_data = Helpers.to_map(npm_package_ref01_data_result)
    assert !npm_package_ref01_data.nil?

    # UPDATE
    npm_package_ref01_data_up0_up = {
    }

    npm_package_ref01_resdata_up0_result = npm_package_ref01_ent.update(npm_package_ref01_data_up0_up, nil)
    npm_package_ref01_resdata_up0 = Helpers.to_map(npm_package_ref01_resdata_up0_result)
    assert !npm_package_ref01_resdata_up0.nil?

    # LOAD
    npm_package_ref01_match_dt0 = {}
    npm_package_ref01_data_dt0_loaded = npm_package_ref01_ent.load(npm_package_ref01_match_dt0, nil)
    assert !npm_package_ref01_data_dt0_loaded.nil?

    # REMOVE
    npm_package_ref01_match_rm0 = {
      "id" => npm_package_ref01_data["id"],
    }
    npm_package_ref01_ent.remove(npm_package_ref01_match_rm0, nil)

  end
end

def npm_package_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "npm_package", "NpmPackageTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = GitlabSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["npm_package01", "npm_package02", "npm_package03", "group01", "group02", "group03", "project01", "project02", "project03", "dist_tag01", "dist_tag02", "dist_tag03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["GITLAB_TEST_NPM_PACKAGE_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "GITLAB_TEST_NPM_PACKAGE_ENTID" => idmap,
    "GITLAB_TEST_LIVE" => "FALSE",
    "GITLAB_TEST_EXPLAIN" => "FALSE",
    "GITLAB_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["GITLAB_TEST_NPM_PACKAGE_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["GITLAB_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["GITLAB_APIKEY"],
      },
      extra || {},
    ])
    client = GitlabSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["GITLAB_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["GITLAB_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/npm_package_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestNpmPackageEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.NpmPackage(nil)
		if ent == nil {
			t.Fatal("expected non-nil NpmPackageEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := npm_packageBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "npm_package." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_NPM_PACKAGE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		npmPackageRef01Ent := client.NpmPackage(nil)
		npmPackageRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "npm_package"}, setup.data), "npm_package_ref01"))
		npmPackageRef01Data["group_id"] = setup.idmap["group01"]
		npmPackageRef01Data["project_id"] = setup.idmap["project01"]

		npmPackageRef01DataResult, err := npmPackageRef01Ent.Create(npmPackageRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		npmPackageRef01Data = core.ToMapAny(npmPackageRef01DataResult)
		if npmPackageRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}

		// UPDATE
		npmPackageRef01DataUp0Up := map[string]any{
		}

		npmPackageRef01ResdataUp0Result, err := npmPackageRef01Ent.Update(npmPackageRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		npmPackageRef01ResdataUp0 := core.ToMapAny(npmPackageRef01ResdataUp0Result)
		if npmPackageRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}

		// LOAD
		npmPackageRef01MatchDt0 := map[string]any{}
		npmPackageRef01DataDt0Loaded, err := npmPackageRef01Ent.Load(npmPackageRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if npmPackageRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

		// REMOVE
		npmPackageRef01MatchRm0 := map[string]any{
			"id": npmPackageRef01Data["id"],
		}
		_, err = npmPackageRef01Ent.Remove(npmPackageRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

	})
}

func npm_packageBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "npm_package", "NpmPackageTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read npm_package test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse npm_package test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"npm_package01", "npm_package02", "npm_package03", "group01", "group02", "group03", "project01", "project02", "project03", "dist_tag01", "dist_tag02", "dist_tag03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_NPM_PACKAGE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_NPM_PACKAGE_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_NPM_PACKAGE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extra,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
